Rebase to flatpak upstream master (commit a641bb76) and set version to 1.4.99

andrunko opened the pull request:
Here are the details on the changes/patches:
- af9aea6e Update Indonesian translation
  - Dropped: translation
- c3df39a0 Update Ukrainian translation
  - Dropped: translation
- 5020790f Suppress debug output when completing
  - Dropped: upstream in 5020790f53f39d34410af669e1793c1ef370d43b
- 26ad9154 Add flatpak in docker seccomp profile
  - Dropped: upstream in 26ad9154c31a484062ffcb46a3818a39ae35a866
- ec2f4372 Checkout using the same mode of the repo
  - Refreshed
- 7eb813b4 build-finish: also export speedwagon files
  - Refreshed
- da748b0f build-finish: Add Endless Discovery Feed Providers
  - Refreshed
- 0bd9ae84 Allow defining an external location for the system installation and use it
  - Refreshed
- 2ad90bf1 Include the external installation directory in XDG_DATA_DIRS
  - Refreshed
- a7342555 flatpak-dir: skip rewriting search provider .ini files for Endless apps
  - Refreshed
- f594e23e common/dir: Change default of core.add-remotes-config-dir to false
  - Refreshed
- 74da5f37 profile/flatpak.sh: add export/bin dirs to PATH
  - Refreshed
- be4fbba2 flatpak-run: Add parental controls support for filtering apps
  - Dropped: Part of e1419b31dba0f45d712952f7adb89ac370f69c79 cherry-picked from pwith's upstream PR
- 7cf2cae5 flatpak-run: Ignore missing app filters for system users
  - Dropped: Part of e1419b31dba0f45d712952f7adb89ac370f69c79 cherry-picked from pwith's upstream PR
- 657f0b1f Revert "Move flatpak_dir_load_appstream_store"
  - Dropped: no longer needed with new parental controls
- f0be6dc2 dir: Support filtering app installs by user’s OARS settings
  - Dropped: Part of 230a8dd802cfbbe135b3231929f784b8641e6439 cherry-picked from pwith's upstream PR
- a129b3fd dir: Allow app filter to disallow app installation entirely
  - Dropped: Part of 230a8dd802cfbbe135b3231929f784b8641e6439 cherry-picked from pwith's upstream PR
- 1eb62556 system-helper: Move parental controls checks to system-helper
  - Dropped: Part of 230a8dd802cfbbe135b3231929f784b8641e6439 cherry-picked from pwith's upstream PR
- 22f79117 system-helper: Don’t try and get parental controls for non-apps/runtimes
  - Dropped: Part of 230a8dd802cfbbe135b3231929f784b8641e6439 cherry-picked from pwith's upstream PR
- 66e44e2c dir: improve error when an app's content is too extreme
  - Dropped: Part of 230a8dd802cfbbe135b3231929f784b8641e6439 cherry-picked from pwith's upstream PR
- cd11a052 doc: Mention parental controls in flatpak-run documentation
  - Dropped: Part of e1419b31dba0f45d712952f7adb89ac370f69c79 cherry-picked from pwith's upstream PR
- 78bdb179 Revert "Stop linking libflatpak against listappstream-glib"
  - Dropped: no longer needed with new parental controls
- a05c6b80 dir: Avoid parental controls check in testlibrary
  - Dropped: Part of 230a8dd802cfbbe135b3231929f784b8641e6439 cherry-picked from pwith's upstream PR
- 2e3abb94 dir: Move find and pull options to innermost scope
  - Dropped: upstream in d18dd69fe42b4d62a9d6ba4bb57d7813787931bd
- f2e3f4ff tests: Consistently use a bash shebang
  - Dropped: upstream in 803fa15b06d910d04a715fd222e968a3d4c78f89
- f25fa7b4 dir: Fix an edge case of resolving collection-refs
  - Dropped: upstream in 8cea78db1f95a9a63a766e0b5c7d9dae68390aad
- d1330170 Made wider use of flatpak_repo_resolve_rev()
  - Dropped: upstream in 253fcc6e35cf882adbee74ebcec28b71718eded2
- 6ff73277 dir: Use the right keyring to verify P2P pulls
  - Dropped: upstream in caedda5b2a0af4dc306313bad8ed26514ea48a45
- 896425c9 dir: Check commit signatures before resolving a ref
  - Dropped: upstream in 915ad583a7bf70e03bf58bf14b9d3bdb7ef33277
- 22dfa485 tests: Add test-p2p-security.sh
  - Dropped: upstream in 6cfad442c5b3f1f4f51301f370fd4bed43d07d68
- fd9c703b common: Properly handle empty arrays of P2P results
  - Dropped: upstream in b8a3075d84322d9f9afa1b152a3283e23615e9a2
- 6358fa7a dir: Do some minor refactoring
  - Dropped: upstream in f67a7ad5bd3a23a0897fc6ab3d9e9360b35e0f9b
- 406ac4c1 Add FLATPAK_ERROR_REF_NOT_FOUND
  - Dropped: upstream in c0d5f1bfc8f09d5c7038422ecd948f26eaad747c
- 10dba0b0 installation: Use FLATPAK_ERROR_REF_NOT_FOUND in one more place
  - Dropped: upstream in c028e0dc1b7861a507c39f76c5bb582ee2150da1
- 073500df common: Fix double free of GError
  - Dropped: upstream in 3164c68e31cf49112d1ff2ce9effe32460c6d97f
- d0a64b82 Add Transifex translation infrastructure
  - Refreshed
- 635bf309 Commit .pot file
  - Dropped: translation
- 9b3b53e3 Pull existing translations from Transifex
  - Dropped: translation
- 71e39f82 Merge new upstream translations
  - Dropped: translation
- 45a2eaf8 common: Do not convert G_IO_ERROR_CANCELLED errors to FLATPAK_ERROR domain when fetching remote metadata.
  - Refreshed
- 5899ab06 common: include ": " between prefix and existing message
  - Refreshed
- b6920fce fixup! profile/flatpak.sh: add export/bin dirs to PATH
  - Squashed
- 843cc0db profile: Fix for openbsd ksh.
  - Dropped: upstream in a48f46f314ae3b4af2e8541143dd9eb5203d6442
- e3058c99 Migrate parental controls support to malcontent
  - Dropped: no longer needed with new parental controls
- afdf72bb Minor compiler warning fix
  - Dropped: Part of 230a8dd802cfbbe135b3231929f784b8641e6439 cherry-picked from pwith's upstream PR
- dd130c90 flatpak-run: Also filter by content type on parental controls checks
  - Dropped: Part of e1419b31dba0f45d712952f7adb89ac370f69c79 cherry-picked from pwith's upstream PR
- 42563c52 system-helper: Apply Parental Controls to privileged users
  - **Refreshed: not part of changes from pwith's upstream PR - is the PR missing this one or is it Endless specific or not longer needed?**
- e08f9c06 parental-controls: Workaround to avoid an additional polkit dialog
  - **Dropped: not part of changes from pwith's upstream PR - is the PR missing this one or is it really not needed anymore?**
- 1238efe1 transaction: Append bundle to correct list on flatpak_transaction_add_install_bundle()
  - Dropped: upstream in 37a7b6edfdb8113ff9d770877f3cd58bf898b8a8
- cf20b428 cli-transaction: Fix some leaks
  - Dropped: upstream in eaec3a8203e5f8dc8b7f2282676981ff6d703110
- 00646377 tests: Temporarily disable failing tests due to recent GLib changes
  - Dropped: no longer needed
- d2d46688 fixup! system-helper: Apply Parental Controls to privileged users
   -  **Dropped: seems to be no longer needed with new parental controls** 
- 664a11ac flatpak-installation: Handle an unconfigured remote when listing remotes
  - Dropped: upstream in 791f31898b629bad5f6e29392be260fb06fb871d
- 25818982 installation: Fix a small leak of a FlatpakCollectionRef
  - Dropped: upstream in f80cce6b8d6a397a232c3a88bf1411f7045cc5e4
- 374b5431 Revert "tests: Temporarily disable failing tests due to recent GLib changes"
  - Dropped: no longer needed
- 59981653 common: Use correct flag for ostree_repo_resolve_rev_ext()
  - Dropped: upstream in 9a258e8a0c36637a43b442592f7d3b5a6eb2ab7c
- 133a7fbe dir: Don't overzealously rewrite an error
  - Dropped: upstream in b44c1a57993905da0a99e5b1f8f1bffd5bb25eee
- 63bfc939 build-bundle: Fix resolving refs
  - Dropped: upstream in 5e65174e8dc8f3853e24b3526a2588d8f4296159

New commits:
- cb329490 Set version to 1.4.99

For parental controls commits I got the patches from https://github.com/pwithnall/flatpak/tree/parental-controls (https://github.com/flatpak/flatpak/pull/2797), squashed the fixup commits and reordered new commits adding error codes to be included before the commits using the error codes.

I also did not include commit https://github.com/pwithnall/flatpak/commit/06d5f50573ed4f083f9046bf22420292917acc0c from the upstream PR as this one is GNOME CI specific and changes point to @pwithnall github repo.

I tagged commit cb329490 with `Version_1.4.99` and the HEAD of the debian branch with `Version_1.4.99_debian`.

https://phabricator.endlessm.com/T26628

andrunko commented:
CI is broken because the new changes depend on ostree >= 2019.2 which is being updated at https://phabricator.endlessm.com/T27607.

andrunko commented:
Regarding the parental controls admin support, I tested with the new package built locally (and also with new ostree) and hit an issue when trying to install `org.gnome.Photos`. When installing I got a polkit dialog (which I should not) asking for permission and after authenticating I got:
```
error: Error deploying: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.freedesktop.Flatpak.override-parental-controls is not registered
error: Failed to install org.gnome.Photos: Error deploying: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.freedesktop.Flatpak.override-parental-controls is not registered
```
This seems to me related to the parental controls patches for admin that got dropped (1eb6255, e08f9c0 and d2d4668) during rebase, but at least part of those are already included in the upstream PR.

@pwithnall could you please have a look and let me know what am I missing here?

pwithnall commented:
> Regarding the parental controls admin support, I tested with the new package built locally (and also with new ostree) and hit an issue when trying to install `org.gnome.Photos`. When installing I got a polkit dialog (which I should not) asking for permission and after authenticating I got:
> 
> ```
> error: Error deploying: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.freedesktop.Flatpak.override-parental-controls is not registered
> error: Failed to install org.gnome.Photos: Error deploying: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.freedesktop.Flatpak.override-parental-controls is not registered
> ```
> 
> This seems to me related to the parental controls patches for admin that got dropped ([1eb6255](https://github.com/endlessm/flatpak/commit/1eb62556185c8e1578422ce111dfeaa9f3d3c43a), [e08f9c0](https://github.com/endlessm/flatpak/commit/e08f9c068726e46d74e38ae92727f8771400e73b) and [d2d4668](https://github.com/endlessm/flatpak/commit/d2d46688a71b2f517426fc13d744515b6fe2fd70)) during rebase, but at least part of those are already included in the upstream PR.

How did you build the package? If you built it but didn’t install it system-wide, the `override-parental-controls` polkit action won’t have been installed somewhere where polkit could see it. Its default rule should allow it (depending on certain conditions), but if that’s not present you’ll get a polkit auth dialogue.

andrunko commented:
> > Regarding the parental controls admin support, I tested with the new package built locally (and also with new ostree) and hit an issue when trying to install `org.gnome.Photos`. When installing I got a polkit dialog (which I should not) asking for permission and after authenticating I got:
> > ```
> > error: Error deploying: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.freedesktop.Flatpak.override-parental-controls is not registered
> > error: Failed to install org.gnome.Photos: Error deploying: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.freedesktop.Flatpak.override-parental-controls is not registered
> > ```
> > 
> > 
> > This seems to me related to the parental controls patches for admin that got dropped ([1eb6255](https://github.com/endlessm/flatpak/commit/1eb62556185c8e1578422ce111dfeaa9f3d3c43a), [e08f9c0](https://github.com/endlessm/flatpak/commit/e08f9c068726e46d74e38ae92727f8771400e73b) and [d2d4668](https://github.com/endlessm/flatpak/commit/d2d46688a71b2f517426fc13d744515b6fe2fd70)) during rebase, but at least part of those are already included in the upstream PR.
> 
> How did you build the package? If you built it but didn’t install it system-wide, the `override-parental-controls` polkit action won’t have been installed somewhere where polkit could see it. Its default rule should allow it (depending on certain conditions), but if that’s not present you’ll get a polkit auth dialogue.

I built it with dpkg-buildpackage locally and installed the new versions, I will check where the polkit files got installed.

andrunko commented:
> > > Regarding the parental controls admin support, I tested with the new package built locally (and also with new ostree) and hit an issue when trying to install `org.gnome.Photos`. When installing I got a polkit dialog (which I should not) asking for permission and after authenticating I got:
> > > ```
> > > error: Error deploying: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.freedesktop.Flatpak.override-parental-controls is not registered
> > > error: Failed to install org.gnome.Photos: Error deploying: GDBus.Error:org.freedesktop.PolicyKit1.Error.Failed: Action org.freedesktop.Flatpak.override-parental-controls is not registered
> > > ```
> > > 
> > > 
> > > This seems to me related to the parental controls patches for admin that got dropped ([1eb6255](https://github.com/endlessm/flatpak/commit/1eb62556185c8e1578422ce111dfeaa9f3d3c43a), [e08f9c0](https://github.com/endlessm/flatpak/commit/e08f9c068726e46d74e38ae92727f8771400e73b) and [d2d4668](https://github.com/endlessm/flatpak/commit/d2d46688a71b2f517426fc13d744515b6fe2fd70)) during rebase, but at least part of those are already included in the upstream PR.
> > 
> > 
> > How did you build the package? If you built it but didn’t install it system-wide, the `override-parental-controls` polkit action won’t have been installed somewhere where polkit could see it. Its default rule should allow it (depending on certain conditions), but if that’s not present you’ll get a polkit auth dialogue.
> 
> I built it with dpkg-buildpackage locally and installed the new versions, I will check where the polkit files got installed.

Just to follow-up here, after restarting `polkit.service` I could install GNOME Photos with `flatpak install` but I still got the polkit dialog. As discussed on slack we could probably merge this and sort out any regression separately.

pwithnall commented:
> Just to follow-up here, after restarting `polkit.service` I could install GNOME Photos with `flatpak install` but I still got the polkit dialog.

Parental controls issues aside, it’s interesting that you needed to restart polkit for it to notice the new rules. It’s supposed to monitor for changes itself and not need restarting (grep for `file_monitor` in the polkit code). That would be worth quickly investigating after merging this.

> As discussed on slack we could probably merge this and sort out any regression separately.

:+1:

andrunko commented:
> > Just to follow-up here, after restarting `polkit.service` I could install GNOME Photos with `flatpak install` but I still got the polkit dialog.
> 
> Parental controls issues aside, it’s interesting that you needed to restart polkit for it to notice the new rules. It’s supposed to monitor for changes itself and not need restarting (grep for `file_monitor` in the polkit code). That would be worth quickly investigating after merging this.

Opened https://phabricator.endlessm.com/T27854 to follow-up on this and the other issues found during testing.

andrunko commented:
> Some comments:
> 
> * Could `common: include ": " between prefix and existing message ` be merged into an earlier downstream commit?
> 
> Apart from that, this looks good to me.

This is now squashed with 1a9e04df487cb17098039f356f8dace076285882, which already touched that code.

I think this is ready now, please let me know if I missed something (apart from the issues to be handled on https://phabricator.endlessm.com/T27854)

andrunko commented:
I updated the branches (source and debian) again to apply debian/patches to the source branch and remove from debian/. Also repushed `Version_*` tags accordingly.

mwleeds commented:
> Usually we push the `Version_x` and `Version_x_debian` at the same time as updating the `master` and `debian-master` branches, with `git push --atomic` but I suppose it won't hurt anything that you've already pushed them.

I see Will said to push them early [here](https://github.com/endlessm/ostree/pull/158#issuecomment-532705908). Never mind.

wjt commented:
Since the `Version_X.Y.Z` tag is meant to match the upstream tag, I don't see any cost in pushing it early. I had always made the `Version_X.Y.Z_debian` tag match the commit where Debian's pristine packaging was imported but I've seen reference to that not being the convention. Personally I like pushing these early because otherwise the CI job will generally fail.

andrunko commented:
@mwleeds Thanks for the review, see comments below.

> We could rebase this on latest master again since there are a few more commits we'll need to backport anyway.
> 
Done.

> "system-helper: Apply Parental Controls to privileged users" does not look correct to me. The version of the patch we have on master doesn't remove handling of `org.freedesktop.Flatpak.app-install` since there's an `else if` to handle it, whereas the rebased version does, and I think that is the reason for the `sudo flatpak install` errors in https://phabricator.endlessm.com/T27854. I don't think we want to merge this and break app installs even for admins.
> 
> We should at least drop that patch, so we don't break app installs, and then in the regression ticket ensure that parental controls still work as expected and apply to admins.
> 
I dropped this patch from the updated branch but the same issues still happen when trying to install using `sudo flatpak install` - I can install the apps fine without sudo though.

I've updated https://phabricator.endlessm.com/T27854 to make sure this patch is reinstated in some form there.

> I didn't review the rest of the parental controls commits in depth; there's a lot there.
> 
> I think "tests: Skip tests that use system helper if uid or gid is 0" can be dropped per the discussion here: [flatpak/flatpak#2999](https://github.com/flatpak/flatpak/pull/2999)
> 
Dropped.

> If we're dropping our downstream translations, how are those going to be re-added? Is that just on @wjt 's pre-release todo list?
> 
Yes, that is how I think it works, @wjt could you please confirm?

andrunko commented:
Note that these changes depend on the [ostree rebase](https://phabricator.endlessm.com/T27607) to be merged first.

wjt commented:
On Fri, 20 Sep, 2019 at 15:08, André Magalhães 
<notifications@github.com> wrote:
>> If we're dropping our downstream translations, how are those going 
>> to be re-added? Is that just on @wjt <https://github.com/wjt> 's 
>> pre-release todo list?
>> 
> Yes, that is how I think it works, @wjt <https://github.com/wjt> 
> could you please confirm?
> 

<https://phabricator.endlessm.com/T27764>

andrunko commented:
bot, retest this please.

andrunko commented:
Thanks everyone for the reviews.
